### ISSUE TITLE
Does it manage arrays and nested objects?

### ISSUE THREAD

**Issue by j2l** [dario-baumberger/obsidian-json-table#4]
Tables are 2D,  JSON are N dimensions and trees with nested data, so I wonder if it's possible to render a flat (2D) table based on JSON more complex objects.

I don't remember the name, but I had something doing this in Google Sheets, adding columns as needed.

**Comment by dario-baumberger** [comment]
Not at the moment. I wanted to implement this functionality, but I never took the time to do it.

I have already given it some thought:

**Example 1:**
```
[{
    "name": "Peter",
    "contact": {
        "mail": "mail@example.com",
        "phone": "blah"
    }
}]
```

should generate: 
```
| name | contact.mail | contact.phone |
| --- | --- | --- |
| Peter | mail@example.com | blah |
```
| name | contact.mail | contact.phone |
| --- | --- | --- |
| Peter | mail@example.com | blah |

**Example 2:**
```
[{
    "name": "Peter",
    "contact": {
        "mail": "mail@example.com",
        "phone": "blah"
    },
    "food": ["Pizza", "Pasta"]
}]
```

should generate: 
```
| name | contact.mail | contact.phone | food[0] | food[1] |
| --- | --- | --- | --- | --- |
| Peter | mail@example.com | blah | Pizza | Pasta |
```
| name | contact.mail | contact.phone | food[0] | food[1] |
| --- | --- | --- | --- | --- |
| Peter | mail@example.com | blah | Pizza | Pasta |

**Example 3:**
```
[{
    "name": "Peter",
    "contact": {
        "mail": "mail@example.com",
        "phone": {
            "private": "blah",
            "work": "foo"
        }
    },
    "food": [
        {
            "type": "Italian",
            "menus": ["Pizza", "Pasta"]
        },
        {
            "type": "Greek",
            "menus": ["Souvlaki", "Tzatziki"]
        }
    ]
}]
 ```

should generate:
```
| name | contact.mail | contact.phone.private | contact.phone.work | food[0].type | food[0].menus[0] | food[0].menus[1] | food[1].type | food[1].menus[0] | food[1].menus[1] |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Peter | mail@example.com | blah | foo | Italian | Pizza | Pasta | Greek | Souvlaki | Tzatziki |
```
| name | contact.mail | contact.phone.private | contact.phone.work | food[0].type | food[0].menus[0] | food[0].menus[1] | food[1].type | food[1].menus[0] | food[1].menus[1] |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Peter | mail@example.com | blah | foo | Italian | Pizza | Pasta | Greek | Souvlaki | Tzatziki |

Does this fit, or is something missing?

Do you have any suggestions for naming the nested contents? Concatenating the keys would be the easiest, and the reference is clear in case two columns have the same name.

I will implement it this way for now. If you have any further inputs, please let me know.

**Comment by j2l** [comment]
It seams that you nailed it! 
No question on my side for now.
Thank you very much for detailing it here.

**Comment by dario-baumberger** [comment]
Perfect, give me a few days. I don't want to make any promises about when I'll be done, but it will come.

**Comment by j2l** [comment]
Take your time to deliver something great!
Have a wonderful week-end.

**Comment by dario-baumberger** [comment]
Could already do a lot. JSON to table was relatively easy compared to table to JSON. There are a few things to consider. I think I could finish at the weekend.

**Comment by j2l** [comment]
> Could already do a lot

Definitely! You already made a great and useful plugin!
Nested JSON is the cherry on the cupcake.

**Comment by dario-baumberger** [comment]
Sorry I didn't manage to finish it over the weekend. It was a bit more complicated than expected.

I've more or less rewritten the whole logic. I'll have another look at everything and, if necessary, split it up a bit further. I also need to add some comments, otherwise I'll have no idea what's going on in six months' time.

**Comment by j2l** [comment]
No problem! Have a great week!

**Comment by kuolemaaa** [comment]
Thank you for your addon. I would ask if something using other addons to render the table is possible.

I see that there are several addons that extends the table functionality and letting you put merged cell (column-wise and row-wise). Some example:
- https://github.com/aidenlx/table-extended
- https://github.com/NicoNekoru/obsidan-advanced-table-xt

To be able to render that json you wrote [here](https://github.com/dario-baumberger/obsidian-json-table/issues/4#issuecomment-2112921054) as this my example:

![example](https://github.com/user-attachments/assets/82e9f450-7643-4cd6-a086-926c2b3027a4)

(i've underlines everything that is a key and not a value)

**Comment by dario-baumberger** [comment]
@j2l I just merged the changes to support arrays and nested objects. I also released version 1.0.3 which includes these changes. 

If you find any bugs or errors, please let me know. :)

**Comment by j2l** [comment]
Merci !
